### PR TITLE
Certs renew use api fix

### LINKED ISSUE #1216: `... certs renew apiserver --use-api` fails: `PEM block type must be CERTIFICATE REQUEST`
## What keywords did you search in kubeadm issues before filing this one?
`certs renew`
 
Found #1190 which is related, but no duplicate.

## Is this a BUG REPORT or FEATURE REQUEST?
BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):

`v1.12.2`

**Environment**:
- **Kubernetes version** (use `kubectl version`): `v1.12.2`
- **Cloud provider or hardware configuration**: HetznerCloud CX31
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04 LTS
- **Kernel** (e.g. `uname -a`): 4.4.0.138.144

## What happened?

Executing `kubeadm alpha phase certs renew apiserver --use-api` fails:
```
failed to renew certificate apiserver: couldn't create certificate signing request:
CertificateSigningRequest.certificates.k8s.io "kubeadm-certccpx4" is invalid:
spec.request: Invalid value: []byte{0x30, 0x82, 0x3, 0x7, […REDACTED…], 0x2,
0xfa, 0xd8, 0x3b, 0x5d, 0x19}: PEM block type must be CERTIFICATE REQUEST
```

## What you expected to happen?

I expected `apiserver.crt` to be renewed based on the current `kubeadm-config` ConfigMap (I updated the `apiServerCertSANs` list to contain another required FQDN).

## How to reproduce it (as minimally and precisely as possible)?

Execute `kubeadm alpha phase certs renew apiserver --use-api`

## Anything else we need to know?

For hints what could be causing this, see also: https://github.com/kubernetes/kubeadm/issues/1190#issuecomment-436080040

### PR DESCRIPTION
**Pull request by liztio**
**What type of PR is this?**
 /kind bug

**What this PR does / why we need it**:
The previous release of kubernetes added a `certs renew` command, with an option to use the [certificates api][certsapi]. This, however, was broken, due to a number of incorrect assumptions and improper testing on my part. 

[certsapi]: https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#approving-certificate-signing-requests

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#1216

**Special notes for your reviewer**:

Follow up to #70772. @neolit123 made the first fix, but that just uncovered more problems. I followed up with another PR to fix this properly. 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/certs/renewal/certsapi.go [hunk 1]
```diff
@@ -130,12 +118,50 @@ func (r *CertsAPIRenewal) Renew(cfg *certutil.Config) (*x509.Certificate, *rsa.P
...
 
-	cert, err := x509.ParseCertificate(req.Status.Certificate)
+	cert, err := certutil.ParseCertsPEM(req.Status.Certificate)
 	if err != nil {
 		return nil, nil, errors.Wrap(err, "couldn't parse issued certificate")
 	}
 
-	return cert, key, nil
+	if len(cert) != 1 {
+		return nil, nil, fmt.Errorf("certificate request %q has %d certificates, wanted exactly 1", req.Name, len(cert))
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: `errors.Errorf()`

File: cmd/kubeadm/app/phases/certs/renewal/certsapi.go [hunk 2]
```diff
@@ -130,12 +118,50 @@ func (r *CertsAPIRenewal) Renew(cfg *certutil.Config) (*x509.Certificate, *rsa.P
...
+	for {
+		select {
+		case ev := <-watcher.ResultChan():
+			switch ev.Type {
+			case watch.Added:
+				continue
+			case watch.Modified:
+				csr, ok := ev.Object.(*certsapi.CertificateSigningRequest)
+				if !ok {
+					return fmt.Errorf("object was of type %T, expected CertificateSigningRequest", ev.Object)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: `errors.Errorf()`

**Review by neolit123 — APPROVED** [review]
thanks @liztio !!
the new make / wait /  parse logic SGTM.
added a couple of `error.Errorf()` comments only.

File: cmd/kubeadm/app/phases/certs/renewal/certsapi.go (hunk 1 above)
Comment (liztio) [reply cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: @neolit123 any particular reason for this? I thought they were pretty much equivalent

Comment (neolit123) [reply cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: @timothysc requested that we use the errors package everywhere.
https://godoc.org/github.com/pkg/errors

File: cmd/kubeadm/app/phases/certs/renewal/certsapi.go [hunk 3]
```diff
@@ -130,12 +118,50 @@ func (r *CertsAPIRenewal) Renew(cfg *certutil.Config) (*x509.Certificate, *rsa.P
...
 
-	return cert, key, nil
+	if len(cert) != 1 {
+		return nil, nil, errors.Errorf("certificate request %q has %d certificates, wanted exactly 1", req.Name, len(cert))
+	}
+
+	return cert[0], key, nil
+}
+
+func waitForReady(name string, csr certstype.CertificateSigningRequestInterface) error {
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: seems weird that this isn't also a cert util, it seems weird that this behavior would on need to live at this location. 

could you add a //TODO to move this to the canonical home for CSRs in 1.14.

File: cmd/kubeadm/app/phases/certs/renewal/certsapi.go [hunk 4]
```diff
@@ -130,12 +118,50 @@ func (r *CertsAPIRenewal) Renew(cfg *certutil.Config) (*x509.Certificate, *rsa.P
...
+	for {
+		select {
+		case ev := <-watcher.ResultChan():
+			switch ev.Type {
+			case watch.Added:
+				continue
+			case watch.Modified:
+				csr, ok := ev.Object.(*certsapi.CertificateSigningRequest)
+				if !ok {
+					return errors.Errorf("object was of type %T, expected CertificateSigningRequest", ev.Object)
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/certs/renewal/certsapi.go]: Did you want to output the data as well as a breadcrumb?

**Review by timothysc — CHANGES_REQUESTED** [review]
minor comments 
/approve

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve

**Pull request merged**